File: typescript/lambda-api-ci/lib/lambda-api-stack.ts
import { LambdaIntegration, MethodLoggingLevel, RestApi } from "aws-cdk-lib/aws-apigateway"
import { PolicyStatement } from "aws-cdk-lib/aws-iam"
import { Function, Runtime, AssetCode, Code } from "aws-cdk-lib/aws-lambda"
import { Duration, Stack, StackProps } from "aws-cdk-lib"
import s3 = require("aws-cdk-lib/aws-s3")
import { Construct } from "constructs"

interface LambdaApiStackProps extends StackProps {
    functionName: string
}

export class CDKExampleLambdaApiStack extends Stack {
    private restApi: RestApi
    private lambdaFunction: Function
    private bucket: s3.Bucket

    constructor(scope: Construct, id: string, props: LambdaApiStackProps) {
        super(scope, id, props)

        this.bucket = new s3.Bucket(this, "WidgetStore")

        this.restApi = new RestApi(this, this.stackName + "RestApi", {
            deployOptions: {
                stageName: "beta",
                metricsEnabled: true,
                loggingLevel: MethodLoggingLevel.INFO,
                dataTraceEnabled: true,
            },
        })

        const lambdaPolicy = new PolicyStatement()
        lambdaPolicy.addActions("s3:ListBucket")
        lambdaPolicy.addActions("s3:getBucketLocation")
        lambdaPolicy.addResources(this.bucket.bucketArn)

        this.lambdaFunction = new Function(this, props.functionName, {
            functionName: props.functionName,
            handler: "handler.handler",
            runtime: Runtime.NODEJS_18_X,
            code: new AssetCode(`./src`),
            memorySize: 512,
            timeout: Duration.seconds(10),
            environment: {
                BUCKET: this.bucket.bucketName,
            },
        })

        this.lambdaFunction.addToRolePolicy(lambdaPolicy)

        this.restApi.root.addMethod("GET", new LambdaIntegration(this.lambdaFunction, {}))
    }
}

File: go/lambda-api-ci/lib/lambdaapistack/lambdaapistack.go
package lambdaapistack

import (
	"fmt"

	"github.com/aws/aws-cdk-go/awscdk"
	"github.com/aws/aws-cdk-go/awscdk/awsapigateway"
	"github.com/aws/aws-cdk-go/awscdk/awsiam"
	"github.com/aws/aws-cdk-go/awscdk/awslambda"
	"github.com/aws/aws-cdk-go/awscdk/awss3"
	"github.com/aws/jsii-runtime-go"
)

type CDKExampleLambdaApiStack awscdk.Stack

type CDKExampleLambdaApiStackProps struct {
	awscdk.StackProps
	FunctionName *string
}

func NewCDKExampleLambdaApiStack(scope awscdk.Construct, id *string, props *CDKExampleLambdaApiStackProps) CDKExampleLambdaApiStack {
	stack := awscdk.NewStack(scope, id, &props.StackProps)

	bucket := awss3.NewBucket(stack, jsii.String("WidgetStore"), nil)

	restApi := awsapigateway.NewRestApi(stack, jsii.String(fmt.Sprintf("%sRestApi", *stack.StackName())), &awsapigateway.RestApiProps{
		DeployOptions: &awsapigateway.StageOptions{
			StageName:        jsii.String("beta"),
			MetricsEnabled:   jsii.Bool(true),
			LoggingLevel:     awsapigateway.MethodLoggingLevel_INFO,
			DataTraceEnabled: jsii.Bool(true),
		},
	})

	lambdaPolicy := awsiam.NewPolicyStatement(&awsiam.PolicyStatementProps{
		Actions:   &[]*string{jsii.String("s3:ListBucket")},
		Resources: &[]*string{bucket.BucketArn()},
	})

	lambdaFunction := awslambda.NewFunction(stack, props.FunctionName, &awslambda.FunctionProps{
		FunctionName: props.FunctionName,
		Handler:      jsii.String("handler.handler"),
		Runtime:      awslambda.Runtime_NODEJS_14_X(),
		Code:         awslambda.NewAssetCode(jsii.String("./src"), nil),
		MemorySize:   jsii.Number(512),
		Timeout:      awscdk.Duration_Seconds(jsii.Number(10)),
		Environment: &map[string]*string{
			"BUCKET": bucket.BucketName(),
		},
		InitialPolicy: &[]awsiam.PolicyStatement{lambdaPolicy},
	})

	restApi.Root().AddMethod(jsii.String("GET"), awsapigateway.NewLambdaIntegration(lambdaFunction, nil), nil)

	return stack
}
